app/ydoc-server-polyglot/src/insight.ts

```diff
@@ -0,0 +1,148 @@
+      })
+      e.removed.forEach((subdoc) => {
+        const subdocProvider = attachedSubdocs.get(subdoc)
+        attachedSubdocs.delete(subdoc)
+        if (subdocProvider != null) {
+          subdocProvider.dispose()
+        }
+      })
+    }
+
+    provider.on('sync', onSync)
+    provider.on('connection-close', onDrop)
+    provider.on('connection-error', onDrop)
+    doc.on('subdocs', onSubdocs)
+
+    function dispose() {
+      provider.disconnect()
+      provider.off('sync', onSync)
+      provider.off('connection-close', onDrop)
+      provider.off('connection-error', onDrop)
+      doc.off('subdocs', onSubdocs)
+      attachedSubdocs.forEach((subdocProvider) => {
+        subdocProvider.dispose()
+      })
+    }
+    return { provider, dispose }
+  }
+
+  const d = new Y.Doc()
+  const project = new DistributedProject(d)
+  console.log(`Insight: got project ${project}`)
+  let provider = null
+  try {
+    provider = attachProvider('ws://localhost:5976/project', 'index', d)
+  } catch (err) {
+    console.error(`Insight: Failed to initialize provider`)
+    console.error(err)
+  }
+  console.log(`Insight: got provider ${provider}`)
+
+  function parseContents(contents: string, syncModule: Ast.MutableModule) {
+    const { code, idMapJson, metadataJson } = splitFileContents(contents)
+    const parsedIdMap = deserializeIdMap(idMapJson)
+
+    const { root, spans } = Ast.parseModuleWithSpans(code, syncModule)
+    syncModule.setRoot(root)
+    spanMap = spans
+
+    Ast.setExternalIds(syncModule, spans, parsedIdMap)
+    console.log(`[parseContents] idMap.entries=${parsedIdMap.entries()}`)
+  }
+
+  console.log(`Insight: listen on sources`)
+  insight.on('source', function (ctx, frame) {
+    console.log(`[source] ctx.uri=${ctx.name}`)
+    if (ctx.name.includes(PROJECT_NAME)) {
+      console.log(`[source] MATCHED ctx=${JSON.stringify(ctx)}`)
+      parseContents(ctx.characters, syncModule)
+    }
+  })
+
+  function resultToString(result: any): string {
+    return '' + result
+  }
+
+  /*
+    class IdMap {
+      private readonly rangeToExpr: Map<SourceRange, ExternalId>
+    }
+    interface SpanMap {
+      nodes: Map<NodeKey, Ast[]>
+      tokens: Map<TokenKey, Token>
+    }
+    type NodeKey = TokenKey = SourceRangeKey
+  */
+  console.log(`Insight: listen on return`)
+  insight.on(
+    'return',
+    function (ctx, frame) {
+      console.log(`[return] ctx='${JSON.stringify(ctx)}'`)
+      console.log(`[return] frame=${Object.getOwnPropertyNames(frame)}`)
+      console.log(`[return] ctx.returnValue='${ctx.returnValue(frame)}'`)
+
+      const p = project
+      const result = ctx.returnValue(frame)
+      if (result && spanMap) {
+        const key = Ast.nodeKey(ctx.charIndex, ctx.charEndIndex - ctx.charIndex)
+        const asts = spanMap.nodes.get(key)
+        if (asts) {
+          const resultValue = resultToString(result)
+          for (const ast of asts) {
+            const editAst = syncModule.getVersion(ast)
```

Comment:
- with #13179 
- we should modify this part to `;(ast as any).fields.get('metadata').set('expressionUpdate', update)`
- e.g. to update the Y.doc server to a value the IDE expects and renders